cpp/src/fil/fil.cu

```diff
@@ -48,6 +48,40 @@ namespace fil {
 
 namespace tl = treelite;
 
+std::ostream& operator<<(std::ostream& os, const cat_sets_owner& cso)
+{
+  os << "\nbits { ";
+  for (uint8_t b : cso.bits)
+    os << std::bitset<8>(b) << " ";
+  os << " }\nmax_matching {";
+  for (int mm : cso.max_matching)
+    os << mm << " ";
+  os << " }";
+  return os;
+}
+
+cat_sets_owner::cat_sets_owner(const std::vector<cat_feature_counters>& cf)
+{
+  max_matching.resize(cf.size());
+  std::size_t bits_size = 0;
+  // feature ID
+  for (std::size_t fid = 0; fid < cf.size(); ++fid) {
+    RAFT_EXPECTS(
+      cf[fid].max_matching >= -1, "@fid %zu: max_matching invalid (%d)", fid, cf[fid].max_matching);
+    RAFT_EXPECTS(cf[fid].n_nodes >= 0, "@fid %zu: n_nodes invalid (%d)", fid, cf[fid].n_nodes);
+
+    max_matching[fid] = cf[fid].max_matching;
+    bits_size +=
+      categorical_sets::sizeof_mask_from_max_matching(max_matching[fid]) * cf[fid].n_nodes;
+
+    RAFT_EXPECTS(bits_size <= INT_MAX,
+                 "@fid %zu: cannot store %zu categories given `int` offsets",
+                 fid,
+                 bits_size);
+  }
+  bits.resize(bits_size);
+}
```

Comment:
I know we previously discussed this function in a different context, and now that I see it _in situ_, I think we don't even need a zip iterator, since we're constructing all of the relevant containers as we go. How about something like the following (composed in Github; excuse any formatting issues)?
```suggestion
template <typename Iter>
cat_sets_owner(Iter cf_begin, Iter cf_end) : 
    max_matching{},
    bits([]() {
      auto bit_size = std::size_t{};
      // Populate max_matching and compute the size of the bits vector along the way.
      // Doing both of these together avoids the overhead of loading the cf values into cache twice.
      std::transform(cf_begin, cf_end, std::back_inserter(max_matching), [](auto& cf_val) {
        auto fid = max_matching.size(); // Tracking feature id for error messages
        RAFT_EXPECTS(
          cf_val.max_matching >= -1, "@fid %zu: max_matching invalid (%d)", fid, cf_val.max_matching);
        RAFT_EXPECTS(cf_val.n_nodes >= 0, "@fid %zu: n_nodes invalid (%d)", fid, cf_val.n_nodes);
        bit_size += categorical_sets::sizeof_mask_from_max_matching(cf_val.max_matching) * cf_val.n_nodes;
        RAFT_EXPECTS(bits_size <= INT_MAX,
             "@fid %zu: cannot store %zu categories given `int` offsets",
             fid,
             bits_size);
        return cf_val.max_matching;
      });
      return bit_size;
    }()) {}
```
We'd obviously want to move this to the header.

I see a few advantages to an approach like this:
1. We are clearly initializing each of our member variables in the initializer list. No chance that `resize` gets called on an uninitialized vector.
2. We remain agnostic to the type of container we are constructing from.
3. Using `std::transform` clearly communicates that we are constructing `max_matching` via a one-to-one transformation from our input elements.

The annoying bit is obviously the fact that we're computing `bit_size` as a side-effect, but if we comment on that to make it clear that it's for performance reasons, I think this should be reasonably clear to future developers.

What do you think? Do we like this?

Reply:
Skipping the indices is neat, but a couple of things are confusing.
If we don't trust this line from cppreference.com, then what do we trust when we read the code? "Before the compound statement that forms the function body of the constructor begins executing, initialization of all direct bases, virtual bases, and non-static data members is finished." AFAIK it's not an experimental feature.

Relying instead on the order of member declaration is IMHO more error-prone, since it's not specified right 
next to it, let alone in the same block. It's also odd that `max_matching` is changed while initializing `bits`. Not a fan of sneaking statements into expressions in a UB-prone context.
In this particular case, it's actually a bug, since I happened to declare `bits` first.
I'll update the code using some of your ideas, maybe it'll become clear enough?

Reply:
Can you open an issue to do a follow up discussion of this point?